core/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java

```diff
@@ -480,32 +462,25 @@ private boolean applyStartedShards(RoutingNodes routingNodes, Iterable<? extends
             } else if (matchingShard.isSameAllocation(startedShard) == false) {
                 logger.debug("{} failed to find shard with matching allocation id in order to start it [failed to find matching shard], ignoring (routing: {}, matched shard routing: {})", startedShard.shardId(), startedShard, matchingShard);
             } else {
-                if (matchingShard.active()) {
+                startedShard = matchingShard;
+                if (startedShard.active()) {
                     logger.trace("{} shard is already started, ignoring (routing: {})", startedShard.shardId(), startedShard);
                 } else {
+                    assert startedShard.initializing();
                     dirty = true;
-                    // override started shard with the latest copy.
-                    startedShard = matchingShard;
-                    routingNodes.started(matchingShard);
+                    routingNodes.started(startedShard);
                     logger.trace("{} marked shard as started (routing: {})", startedShard.shardId(), startedShard);
-                }
-            }
-
-            // startedShard is the current state of the shard (post relocation for example)
-            // this means that after relocation, the state will be started and the currentNodeId will be
-            // the node we relocated to
-            if (startedShard.relocatingNodeId() == null) {
-                continue;
-            }
 
-            RoutingNodes.RoutingNodeIterator sourceRoutingNode = routingNodes.routingNodeIter(startedShard.relocatingNodeId());
-            if (sourceRoutingNode != null) {
-                while (sourceRoutingNode.hasNext()) {
-                    ShardRouting shard = sourceRoutingNode.next();
-                    if (shard.isRelocationSourceOf(startedShard)) {
-                        dirty = true;
-                        sourceRoutingNode.remove();
-                        break;
+                    if (startedShard.relocatingNodeId() != null) {
+                        // relocation target has been started, remove relocation source
+                        RoutingNode relocationSourceNode = routingNodes.node(startedShard.relocatingNodeId());
+                        if (relocationSourceNode != null) {
```

Comment:
I cannot think of a situation where it is null. At that point in the method, startedShard is the existing entry in the routing table. So it should also have the corresponding entry for the relocation source. I've omitted the null checks...